src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1978,183 +1987,205 @@ BuildJobTreeTaskList(Job *jobTree)
-	uint32 rangeTableIndex = 0;
-	const uint32 fragmentSize = sizeof(RangeTableFragment);
-	uint64 largestTableSize = 0;
-
-	/* find filters on partition columns */
-	ExtractQueryWalker((Node *) subquery, &queryList);
-	foreach(queryCell, queryList)
-	{
-		Query *query = (Query *) lfirst(queryCell);
-		bool leafQuery = LeafQuery(query);
-
-		if (!leafQuery)
-		{
-			continue;
-		}
-
-		/* we have some filters on partition column */
-		opExpressionList = PartitionColumnOpExpressionList(query);
-		if (opExpressionList != NIL)
-		{
-			break;
-		}
-	}
+	Oid relationId = 0;
+	int shardCount = 0;
+	int shardOffset = 0;
+	DistTableCacheEntry *targetCacheEntry = NULL;
+	RelationRestrictionContext *relationRestrictionContext =
+		plannerRestrictionContext->relationRestrictionContext;
+	bool allRelationsJoinedOnPartitionKey =
+		AllRelationsJoinedOnPartitionKey(plannerRestrictionContext);
 
 	/* get list of all range tables in subquery tree */
 	ExtractRangeTableRelationWalker((Node *) subquery, &rangeTableList);
 
 	/*
-	 * For each range table entry, first we prune shards for the relation
-	 * referenced in the range table. Then we sort remaining shards and create
-	 * fragments in this order and add these fragments to fragment combination
-	 * list.
+	 * Find the first relation that is not a reference table. We'll use the shards
+	 * of that relation as the target shards.
 	 */
 	foreach(rangeTableCell, rangeTableList)
 	{
 		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
-		Oid relationId = rangeTableEntry->relid;
-		List *shardIntervalList = LoadShardIntervalList(relationId);
-		List *finalShardIntervalList = NIL;
-		ListCell *fragmentCombinationCell = NULL;
-		ListCell *shardIntervalCell = NULL;
-		uint32 tableId = rangeTableIndex + 1; /* tableId starts from 1 */
-		uint32 finalShardCount = 0;
-		uint64 tableSize = 0;
+		DistTableCacheEntry *cacheEntry = NULL;
 
-		if (opExpressionList != NIL)
-		{
-			Var *partitionColumn = PartitionColumn(relationId, tableId);
-			List *whereClauseList = ReplaceColumnsInOpExpressionList(opExpressionList,
-																	 partitionColumn);
-			finalShardIntervalList = PruneShardList(relationId, tableId, whereClauseList,
-													shardIntervalList);
-		}
-		else
+		relationId = rangeTableEntry->relid;
+		cacheEntry = DistributedTableCacheEntry(relationId);
+
+		if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
 		{
-			finalShardIntervalList = shardIntervalList;
+			continue;
 		}
 
-		/* if all shards are pruned away, we return an empty task list */
-		finalShardCount = list_length(finalShardIntervalList);
-		if (finalShardCount == 0)
+		if (targetShardIntervalList == NIL)
 		{
-			return NIL;
+			targetCacheEntry = DistributedTableCacheEntry(relationId);
+			shardCount = targetCacheEntry->shardIntervalArrayLength;
+			break;
 		}
+	}
 
-		fragmentCombinationCell = list_head(fragmentCombinationList);
+	for (shardOffset = 0; shardOffset < shardCount; shardOffset++)
```

Comment:
Can we move next line from above to here before `for` loop?
`shardCount = targetCacheEntry->shardIntervalArrayLength;

Also, maybe add an `Assert(targetCacheEntry != NULL)` before it.